indexer/Worker.cc

```diff
@@ -36,8 +54,424 @@ struct MessageQueuePair {
+    this->sema = &S;
+  }
+
+  void ForgetSema() override {
+    this->sema = nullptr;
+  }
+};
+
+class IndexerFrontendAction : public clang::ASTFrontendAction {
+  const IndexerPPConfig &config;
+  SemanticAnalysisJobResult &result;
+  // ^ These fields are just for passing down information
+  // down to the IndexerPPCallbacks value.
+public:
+  IndexerFrontendAction(const IndexerPPConfig &config,
+                        SemanticAnalysisJobResult &result)
+      : config(config), result(result) {}
+
+  bool usesPreprocessorOnly() const override {
+    return false;
+  }
+
+  std::unique_ptr<clang::ASTConsumer>
+  CreateASTConsumer(clang::CompilerInstance &compilerInstance,
+                    llvm::StringRef filepath) override {
+    auto &preprocessor = compilerInstance.getPreprocessor();
+    preprocessor.addPPCallbacks(std::make_unique<IndexerPPCallbacks>(
+        compilerInstance.getSourceManager(), config, result));
+    return std::make_unique<IndexerASTConsumer>(compilerInstance, filepath);
+  }
+};
+
+class IndexerFrontendActionFactory
+    : public clang::tooling::FrontendActionFactory {
+  const IndexerPPConfig &config;
+  SemanticAnalysisJobResult &result;
+
+public:
+  IndexerFrontendActionFactory(const IndexerPPConfig &config,
+                               SemanticAnalysisJobResult &result)
+      : config(config), result(result) {}
+  // /// Invokes the compiler with a FrontendAction created by create().
+  // bool runInvocation(std::shared_ptr<CompilerInvocation> Invocation,
+  //                    FileManager *Files,
+  //                    std::shared_ptr<PCHContainerOperations> PCHContainerOps,
+  //                    DiagnosticConsumer *DiagConsumer) override;
+
+  virtual std::unique_ptr<clang::FrontendAction> create() override {
+    return std::make_unique<IndexerFrontendAction>(config, result);
+  }
+};
+
+class IndexerDiagnosticConsumer : public clang::DiagnosticConsumer {
+public:
+  void HandleDiagnostic(clang::DiagnosticsEngine::Level level,
+                        const clang::Diagnostic &info) override {
+    // Just dropping all diagnostics on the floor for now.
+    // FIXME(def: surface-diagnostics)
+  }
+};
+
 } // namespace
 
+static SemanticAnalysisJobResult
+performSemanticAnalysis(SemanticAnalysisJobDetails &&job) {
+  clang::FileSystemOptions fileSystemOptions;
+  fileSystemOptions.WorkingDir = std::move(job.command.Directory);
+
+  llvm::IntrusiveRefCntPtr<clang::FileManager> fileManager(
+      new clang::FileManager(fileSystemOptions, nullptr));
+
+  auto args = std::move(job.command.CommandLine);
+  args.push_back("-fsyntax-only");   // Only type-checking, no codegen.
+  args.push_back("-Wno-everything"); // Warnings aren't helpful.
+  // Should we add a CLI flag to pass through extra arguments here?
+
+  auto projectRoot = std::filesystem::current_path().string();
+  IndexerPPConfig config{
+      AbsolutePath::tryFrom(std::string_view(projectRoot)).value(),
+      HeaderFilter(), false};
```

Comment:
TODO: Add CLI flags and pass these from the driver to the worker.

Reply:
Related TODO: The determinism flag should probably also make the scheduler deterministic. That shouldn't be part of this patch, but it should have an associated FIXME and tracking issue.